contracts/protocol/modules/v1/AaveV3LeverageModule.sol

```diff
@@ -0,0 +1,1147 @@
+     */
+    function _validateNewCollateralAsset(ISetToken _setToken, IERC20 _asset) internal view {
+        require(!collateralAssetEnabled[_setToken][_asset], "Collateral already enabled");
+
+        (address aToken, , ) = protocolDataProvider.getReserveTokensAddresses(address(_asset));
+        require(address(underlyingToReserveTokens[_asset].aToken) == aToken, "Invalid aToken address");
+
+        ( , , , , , bool usageAsCollateralEnabled, , , bool isActive, bool isFrozen) = protocolDataProvider.getReserveConfigurationData(address(_asset));
+        // An active reserve is an alias for a valid reserve on Aave.
+        // We are checking for the availability of the reserve directly on Aave rather than checking our internal `underlyingToReserveTokens` mappings,
+        // because our mappings can be out-of-date if a new reserve is added to Aave
+        require(isActive, "IAR");
+        // A frozen reserve doesn't allow any new deposit, borrow or rate swap but allows repayments, liquidations and withdrawals
+        require(!isFrozen, "FAR");
+        require(usageAsCollateralEnabled, "CNE");
+    }
+
+    /**
+     * @dev Validates if a new asset can be added as borrow asset for given SetToken
+     */
+    function _validateNewBorrowAsset(ISetToken _setToken, IERC20 _asset) internal view {
+        require(!borrowAssetEnabled[_setToken][_asset], "Borrow already enabled");
+
+        ( , , address variableDebtToken) = protocolDataProvider.getReserveTokensAddresses(address(_asset));
+        require(address(underlyingToReserveTokens[_asset].variableDebtToken) == variableDebtToken, "Invalid variable debt token address");
+
+        (, , , , , , bool borrowingEnabled, , bool isActive, bool isFrozen) = protocolDataProvider.getReserveConfigurationData(address(_asset));
+        require(isActive, "IAR");
+        require(!isFrozen, "FAR");
+        require(borrowingEnabled, "BNE");
+    }
+
+    /**
+     * @dev Reads aToken balance and calculates default position unit for given collateral aToken and SetToken
+     *
+     * @return uint256       default collateral position unit
+     */
+    function _getCollateralPosition(ISetToken _setToken, IAToken _aToken, uint256 _setTotalSupply) internal view returns (uint256) {
+        uint256 collateralNotionalBalance = _aToken.balanceOf(address(_setToken));
+        return collateralNotionalBalance.preciseDiv(_setTotalSupply);
+    }
+
+    /**
+     * @dev Reads variableDebtToken balance and calculates external position unit for given borrow asset and SetToken
+     *
+     * @return int256       external borrow position unit
+     */
+    function _getBorrowPosition(ISetToken _setToken, IERC20 _borrowAsset, uint256 _setTotalSupply) internal view returns (int256) {
+        uint256 borrowNotionalBalance = underlyingToReserveTokens[_borrowAsset].variableDebtToken.balanceOf(address(_setToken));
+        return borrowNotionalBalance.preciseDivCeil(_setTotalSupply).toInt256().mul(-1);
+    }
+}
```

Comment:
fantastic natspec and comments here 💯

Reply:
Thank you. They are almost entirely copied from the original aaveLeverageModule so the compliments should probably go to that codes author😅